---
title: "analysis"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
This script provides all the code used for the analysis of the publication entitled "Epigenetic signatures of social status in female free-ranging spotted hyenas (Crocuta crocuta)" by Vullioud et al. 2024. 
The code using the "path_to_annotation_file" argument cannot be run as such. The full dataset is available on NCBI BioProject ID PRJNA1036526, including reference genome "crocuta.fasta". Novel annotation file and read count tables are available here https://doi.org/10.5061/dryad.m0cfxpp9b78. Metadata are listed in Supplementary Material. Please contact the corresponding author for the annotation file and the full data set. The dataseet is pretty big. If you like to try the first part of the code, a smaller dataset (data(full_300_ws_tables_1000) is provided here, which consists of the first 1000 windows. 


```{r setup, echo = FALSE, eval = FALSE}
library(Weyrich23)
library(dplyr)
```

## fit Negative-binomial model on full data set and subsets

```{r, find DMRs,  include =TRUE, eval = FALSE}
data("full_300_ws_tables_1000") ## for access to the full dataset please contact the corresponding author. 

set.seed(123) 
describes_samples(full_300_ws_tables_1000) 


## create sets and subsets
model_1_one_fold <- prepare_balanced_k_set(k = 1, variable_to_balance = "group", tables =full_300_ws_tables)
model_1_six_fold <- prepare_balanced_k_set(k = 6, variable_to_balance = "group", tables =full_300_ws_tables)

## fit negative binomial model on the windows
model_1_one_fold <- add_DMRs_to_set(model_1_one_fold, 
                                    ull_300_ws_tables, formula = "~ group + age_group", #
                                    p_cut = 0.001) ##the cut on the FDR is done later
# can change the fitting procedure. PQL is slightly more conservative but not any better in cross-validation. 

## extract indexes
index <- dplyr::bind_cols(model_1_one_fold$DMR_train) %>% 
  dplyr::filter(coef == "groupLR" & FDR <= 0.05) %>% 
  dplyr::pull(index) ## extract indexes of DMRS windows with FDR < 0.05

## fit negative binomial model on the windows in the subsets
model_1_six_fold <- add_DMRs_to_set(model_1_six_fold, full_300_ws_tables, formula = "~ group + age_group", p_cut = 0.001, index = index) ## keep the result for indexes
```


## Part 1 DMR identification and DMR validation

```{r, clean DMRs,  include =TRUE, warning=FALSE,  eval = FALSE}
data("model_1_one_fold") ## results obtained with the full dataset (= analysis)
data("model_1_six_fold") ## results obtained with the full subsets (= analysis)
data("loc_table") 
path_to_annotation_table <- system.file("extdata", "crocuta_liftoff_Hhy_ASM300989v1_addPromoter.gtf", package = "Weyrich23")

DMR_group <- create_DMR_group_table(model_1_one_fold, model_1_six_fold, path_to_annotation_table, loc_table)

## descriptive stats given in the manuscript 
## numbers given in text (results + abstract)
## get basic stats
length(unique(DMR_group$block)) ## n unique DMRs
length(unique(DMR_group$block[DMR_group$annot_null])) # n not annotated DMRs
length(unique(DMR_group$block[!DMR_group$annot_null])) # m annotated DMRs
length(unique(DMR_group$block[!DMR_group$annot_null & DMR_group$coherent_block])) # n coherent and annotated DMR
length(unique(DMR_group$block[DMR_group$coherent_block])) # n coherent DMR
length(unique(DMR_group$block[!DMR_group$coherent_block])) # n incoherent DMR

## get the total number of annotated regions (gene, intron, exon, TSS, promoter, etc.)
annotated_DMR_group <- create_annot_DMRs(DMR_group)
length(unique(annotated_DMR_group$annot.gene_id)) ## total number of genes

annot_type <- get_annotation_type(annotated_DMR_group)
table(annot_type$annot.type)

fig1(DMR_group)
fig2(DMR_group)
```

## Part 2 mitochondria genes (mito genes) vs nuclear genes (non-mito genes) 

```{r, mitochondria,  include =TRUE, warning=FALSE, eval=FALSE} 
mito_table_DNA <- compute_mean_cpm_mito(full_300_ws_tables, path_to_annotation_table) ## big files. you can access the computed table stored in data
mito_table_RNA <- compute_mean_cpm_mito(RNA_table)
# data(mito_table_DNA) ## computed table available with this package
# data(mito_table_RNA) ## computed table available with this package


mito_table_RNA %>%
  group_by(group) %>% 
  summarise(n = length(unique(ID))) ## get n = for RNA data

## get simple p-value for delta between mito vs non-mito and HR vs LR
coin::wilcoxsign_test(mean_count_TMM ~ as.factor(type_of_genes), data = mito_table_DNA, distribution = "exact")  
coin::wilcoxsign_test(mean_count_TMM ~ as.factor(type_of_genes), data = mito_table_RNA, distribution = "exact") ## paired Wilocoxon for the mito vs non-mito genes #since the samples are the same

coin::wilcox_test(mean_count_TMM ~ as.factor(group), data = mito_table_DNA, distribution = "exact") 
coin::wilcox_test(mean_count_TMM ~ as.factor(group), data = mito_table_RNA, distribution = "exact") 
## non-paired Wilocoxon for the mito vs non-mito genes since the samples are different

### box plot
fig3(mito_table_DNA, mito_table_RNA) 

```

## Part 3 Random Forest (RF)

```{r, rf,  include =TRUE, warning=FALSE, eval = FALSE}
data(model_1_one_fold)
data(DMR_group) 
load("data-raw/full_300_ws_tables.rda")

set.seed(123)
oob_result <- estimate_oob(DMR_group, model_1_one_fold, full_300_ws_tables, 1000) ### loop on the function to have less variation due to the random nature of the analysis. // only matter for the small n. for the random draw (table in the supplementary material (Vullioud et al. 2024)
```

## supplementary material
```{r, sup,  include =TRUE, warning=FALSE, eval = FALSE}
data(full_300_ws_tables_1000)
fig_s2b(full_300_ws_tables_1000, save = FALSE)

## comparison Young/adult fitted separately 
data("DMR_group")
data("DMR_adult_young_comp")
data("DMR_bootstrap_adult_young")
fig_s4(DMR_group, DMR_adult_young_comp, DMR_bootstrap_adult_young, save = FALSE)
```
